src/libstd/sys_common/backtrace.rs

```diff
@@ -127,13 +127,27 @@ fn filter_frames(frames: &[Frame],
     (skipped_before, skipped_after)
 }
 
+static MARK_START: fn(&mut FnMut()) = mark_start;
 
-/// Fixed frame used to clean the backtrace with `RUST_BACKTRACE=1`.
+/// Fixed frame used to clean the backtrace with `RUST_BACKTRACE=1`
 #[inline(never)]
-pub fn __rust_begin_short_backtrace<F, T>(f: F) -> T
-    where F: FnOnce() -> T, F: Send, T: Send
-{
-    f()
+fn mark_start(f: &mut FnMut()) {
+    f();
+    unsafe {
+        asm!("" ::: "memory" : "volatile"); // A dummy statement to prevent tail call optimization
```

Comment:
This may not compile on platforms like asm.js, so I wonder if we could take a perhaps different route to implement this? Could this call a `#[inline(never)]` function which doesn't throw using the `notail` attribute on the call? I think that'd do the same thing, right?

Reply:
`f` may throw here, so we cannot use `notail`. We need something that is opaque to LLVM. I suggest we just disable this `asm!` expressions on platforms which do not support it.

Reply:
Ah yes indeed that's why we can't notail the call to `f`, but I"m thinking something like:

```rust
fn mark_start(f: &mut FnMut()) {
    f();
    #[notail]
    foo();

    #[inline(never)]
    fn foo() {}
}
```

Reply:
LLVM could still see that `foo` has no side effect, and just remove the call.

Reply:
Ah indeed, in that case could something more portable than `asm!` be used, like [volatile reads](https://play.rust-lang.org/?gist=cddccf7488f9953c5f4065e9f469e2cc&version=nightly) or something like that?

Reply:
I don't think volatile operations will work as they could be reordered to happen before the call to `f`, if `f` contains no volatile operations itself.

Reply:
Does LLVM accept empty inline assembly for asm.js?

Reply:
I was asking about LLVM not rustc though. I do not know if the reordering could happen given LLVM's current optimizations.

Reply:
At least in the tests I was running it wasn't reordering, but I'm not sure if that's a guarantee? I figure we could at least try it as this'll need to be upgraded for emscripten anyway.